File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int>lmaxi(n+1,0);
        vector<int>rmaxi(n+1,0);

        int maxi = INT_MIN;

        for(int i=0;i<n;i++){
            maxi = max(maxi,height[i]);
            lmaxi[i]=maxi;
        }

        maxi = INT_MIN;

        for(int j=n-1;j>=0;j--){
            maxi = max(maxi,height[j]);
            rmaxi[j]=maxi;
        }
        int ans = 0;
        
        for(int i=0;i<n;i++){
            int waterSt = min(lmaxi[i],rmaxi[i])-height[i];
            ans+=waterSt;
        }
        return ans;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] ht) {
        int n = ht.length;

        int []lmaxi = new int[n];
        int []rmaxi = new int[n];

        int lMaxiNum = -1;
        int rMaxiNum = -1;

        for(int i=0;i<n;i++){
            lMaxiNum = Math.max(lMaxiNum,ht[i]);
            lmaxi[i]=lMaxiNum;
        }

        for(int i=n-1;i>=0;i--){
            rMaxiNum = Math.max(rMaxiNum,ht[i]);
            rmaxi[i]=rMaxiNum;
        }

        int totWtr = 0;

        for(int i=0;i<n;i++){
            int h = Math.min(lmaxi[i],rmaxi[i]);
            int currWtr = h - ht[i];
            totWtr+=currWtr;
        }
        return totWtr;
    }
}